### PR TITLE
 Combine Free Queue and WASM Helper Ring Buffer Implementations

### PR DESCRIPTION
**Pull request by MeenuyD**
Combine Free Queue and WASM Helper Ring Buffer Implementations

Objective:
The objective of this issue is to combine two different implementations of a ring buffer into a single, cohesive implementation that incorporates the features and advantages of both the Free Queue and WASM Helper approaches.

Description:
Currently, we have two separate implementations of a ring buffer:

Free Queue: This implementation provides a single-producer/single-consumer lock-free FIFO backed by SharedArrayBuffer. It is optimized for managing audio data in a multi-threaded environment.
WASM Helper: This implementation utilizes WebAssembly (WASM) to optimize the ring buffer for single-threaded operation. It is designed to efficiently handle audio data in a single-threaded environment.

#347

### CODE REVIEW

**Comment by Daksh-10** [comment]
@MeenuyD you didn't signed the CLA, then only it will be allowed to merged. I am also learning and working on the same project.

**Comment by hoch** [comment]
IIUC this only adds a new file to the `lib` directory. I believe other files (the user of this library) will also need to be updated, and the redundant files should be removed.

**Comment by MeenuyD** [comment]
Yes I will do it as soon as possible

File: src/audio-worklet/design-pattern/lib/ring-buffer.js
Comment (hoch) [inline src/audio-worklet/design-pattern/lib/ring-buffer.js:1]: The convention is to use the class name for the file name: `FreeQueue.js`

File: src/audio-worklet/design-pattern/lib/ring-buffer.js [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+    Atomics.store(this.states, this.States.READ, 0);
+    Atomics.store(this.states, this.States.WRITE, 0);
+  }
+}
+
+/**
+ * Improved RingBuffer implementation for audio worklets.
+ * This implementation is optimized for single-threaded operation.
+ */
+class RingBuffer {
```
Comment (hoch) [inline src/audio-worklet/design-pattern/lib/ring-buffer.js:205]: The goal of this PR is to remove this class. We're unifying everything into one class `FreeQueue`.

File: src/audio-worklet/design-pattern/wasm-ring-buffer/ring-buffer-worklet-processor.js [hunk 2]
```diff
@@ -41,16 +42,14 @@ class RingBufferWorkletProcessor extends AudioWorkletProcessor {
     this._channelCount = options.processorOptions.channelCount;
 
     // RingBuffers for input and output.
-    this._inputRingBuffer =
-        new RingBuffer(this._kernelBufferSize, this._channelCount);
-    this._outputRingBuffer =
-        new RingBuffer(this._kernelBufferSize, this._channelCount);
+    this._buffer = new FreeQueue(this._kernelBufferSize, this._channelCount);
```
Comment (hoch) [inline src/audio-worklet/design-pattern/wasm-ring-buffer/ring-buffer-worklet-processor.js:45]: `_freeQueue` would be a better variable name.

File: src/audio-worklet/design-pattern/wasm-ring-buffer/ring-buffer-worklet-processor.js [hunk 3]
```diff
@@ -41,16 +42,14 @@ class RingBufferWorkletProcessor extends AudioWorkletProcessor {
...
 
     // For WASM memory, also for input and output.
     this._heapInputBuffer =
-        new HeapAudioBuffer(Module, this._kernelBufferSize, this._channelCount);
+      new Float32Array(Module.instance.exports.memory.buffer, 0, this._kernelBufferSize * this._channelCount);
```
Comment (hoch) [inline src/audio-worklet/design-pattern/wasm-ring-buffer/ring-buffer-worklet-processor.js:49]: So we're replacing `HeapAudiobuffer` by providing the Module memory with the address offset? Just to confirm.

nit: the overflowing code uses 4-space indent - See https://google.github.io/styleguide/jsguide.html#formatting-indent.

File: src/audio-worklet/design-pattern/wasm-ring-buffer/ring-buffer-worklet-processor.js [hunk 4]
```diff
@@ -41,16 +42,14 @@ class RingBufferWorkletProcessor extends AudioWorkletProcessor {
...
     this._heapOutputBuffer =
-        new HeapAudioBuffer(Module, this._kernelBufferSize, this._channelCount);
+      new Float32Array(Module.instance.exports.memory.buffer, this._kernelBufferSize * this._channelCount, this._kernelBufferSize * this._channelCount);
```
Comment (hoch) [inline src/audio-worklet/design-pattern/wasm-ring-buffer/ring-buffer-worklet-processor.js:51]: Ditto.

nit: Line wrap at col 80.

File: src/audio-worklet/design-pattern/wasm-ring-buffer/ring-buffer-worklet-processor.js [hunk 5]
```diff
@@ -72,29 +71,36 @@ class RingBufferWorkletProcessor extends AudioWorkletProcessor {
...
-      this._inputRingBuffer.pull(this._heapInputBuffer.getChannelData());
+    if (this._buffer.isFrameAvailable(this._kernelBufferSize)) {
+      // Pull the queued data from the combined buffer.
+      this._buffer.pull(this._heapInputBuffer, this._kernelBufferSize);
+
+      // Process the data using the kernel.
+      this._kernel.process(
+        this._heapInputBuffer.byteOffset,
+        this._heapOutputBuffer.byteOffset,
+        this._channelCount
```
Comment (hoch) [inline src/audio-worklet/design-pattern/wasm-ring-buffer/ring-buffer-worklet-processor.js:85]: nit: close the parenthesis at the same line here.

File: src/audio-worklet/design-pattern/wasm-ring-buffer/ring-buffer-worklet-processor.js [hunk 6]
```diff
@@ -72,29 +71,36 @@ class RingBufferWorkletProcessor extends AudioWorkletProcessor {
...
+      );
+
+      // Push the processed data back into the combined buffer.
+      this._buffer.push(this._heapOutputBuffer, this._kernelBufferSize);
+    }
+    // if (this._inputRingBuffer.framesAvailable >= this._kernelBufferSize) {
+    //   // Get the queued data from the input ring buffer.
+    //   this._inputRingBuffer.pull(this._heapInputBuffer.getChannelData());
```
Comment (hoch) [inline src/audio-worklet/design-pattern/wasm-ring-buffer/ring-buffer-worklet-processor.js:93]: L92: Uncomment the commented line.
L93: Why is this line commented out?

File: src/audio-worklet/design-pattern/wasm-supersaw/synth-processor.js [hunk 7]
```diff
@@ -16,7 +16,8 @@ class SynthProcessor extends AudioWorkletProcessor {
     // Create an instance of Synthesizer and WASM memory helper. Then set up an
     // event handler for MIDI data from the main thread.
     this._synth = new Module.Synthesizer(sampleRate);
-    this._wasmHeapBuffer = new HeapAudioBuffer(Module, NUM_FRAMES, 1, 1);
+    // this._wasmHeapBuffer = new HeapAudioBuffer(Module, NUM_FRAMES, 1, 1);
```
Comment (hoch) [inline src/audio-worklet/design-pattern/wasm-supersaw/synth-processor.js:19]: This is commented out.

File: src/audio-worklet/design-pattern/wasm-supersaw/synth-processor.js [hunk 8]
```diff
@@ -16,7 +16,8 @@ class SynthProcessor extends AudioWorkletProcessor {
...
+    this._combinedBuffer = new FreeQueue(NUM_FRAMES, 1);
```
Comment (hoch) [inline src/audio-worklet/design-pattern/wasm-supersaw/synth-processor.js:20]: `_freeQueue` would be a better variable name.

File: src/audio-worklet/design-pattern/wasm-supersaw/synth-processor.js [hunk 9]
```diff
@@ -27,8 +28,10 @@ class SynthProcessor extends AudioWorkletProcessor {
     // Call the render function to write into the WASM buffer. Then clone the
     // rendered data in the first channel to process() callback's output
     // buffer.
-    this._synth.render(this._wasmHeapBuffer.getHeapAddress(), NUM_FRAMES);
-    outputBuffer.set(this._wasmHeapBuffer.getChannelData(0));
+    // this._synth.render(this._wasmHeapBuffer.getHeapAddress(), NUM_FRAMES);
+    // outputBuffer.set(this._wasmHeapBuffer.getChannelData(0));
```
Comment (hoch) [inline src/audio-worklet/design-pattern/wasm-supersaw/synth-processor.js:32]: These two lines are commented out.

File: src/audio-worklet/design-pattern/wasm/wasm-worklet-processor.js [hunk 10]
```diff
@@ -3,8 +3,7 @@
 // found in the LICENSE file.
 
 import Module from './simple-kernel.wasmmodule.js';
-import {RENDER_QUANTUM_FRAMES, MAX_CHANNEL_COUNT, HeapAudioBuffer}
-  from '../lib/wasm-audio-helper.js';
+import {RENDER_QUANTUM_FRAMES, MAX_CHANNEL_COUNT,FreeQueue} from '../lib/ring-buffer.js';
```
Comment (hoch) [inline src/audio-worklet/design-pattern/wasm/wasm-worklet-processor.js:6]: nit: Line wrap at 80
nit: add a space after comma

File: src/audio-worklet/design-pattern/wasm/wasm-worklet-processor.js [hunk 11]
```diff
@@ -21,10 +20,7 @@ class WASMWorkletProcessor extends AudioWorkletProcessor {
 
     // Allocate the buffer for the heap access. Start with stereo, but it can
     // be expanded up to 32 channels.
-    this._heapInputBuffer = new HeapAudioBuffer(
-        Module, RENDER_QUANTUM_FRAMES, 2, MAX_CHANNEL_COUNT);
-    this._heapOutputBuffer = new HeapAudioBuffer(
-        Module, RENDER_QUANTUM_FRAMES, 2, MAX_CHANNEL_COUNT);
+    this._buffer = new FreeQueue(RENDER_QUANTUM_FRAMES, MAX_CHANNEL_COUNT);
```
Comment (hoch) [inline src/audio-worklet/design-pattern/wasm/wasm-worklet-processor.js:23]: Ditto on the variable name.

**Review by hoch — COMMENTED** [review]
Thanks for your contribution! I added some comments.

File: src/audio-worklet/design-pattern/wasm-ring-buffer/ring-buffer-worklet-processor.js (hunk 3 above)
Comment (MeenuyD) [reply src/audio-worklet/design-pattern/wasm-ring-buffer/ring-buffer-worklet-processor.js:49]: Yes

**Comment by hoch** [comment]
I am closing this because the similar task is completed by https://github.com/GoogleChromeLabs/web-audio-samples/pull/369.

**Pull request closed without merging**